### PR TITLE
feat(channel): Make channel feature additive

### PR DESCRIPTION
**Pull request by tottoto**
## Motivation

Allows users to use transport feature independent from channel.

## Solution

> [!WARNING] 
> This proposal includes a breaking change.

Makes channel feature additive.

### CODE REVIEW

**Comment by tottoto** [comment]
The failed ci would be fixed by #1628.

File: examples/Cargo.toml [hunk 1]
```diff
@@ -298,13 +298,13 @@ hyper-warp-multiplex = ["hyper-warp"]
...
 mock = ["tokio-stream", "dep:tower"]
-tower = ["dep:hyper", "dep:tower", "dep:http"]
+tower = ["dep:hyper", "dep:tower", "tower?/timeout", "dep:http"]
 json-codec = ["dep:serde", "dep:serde_json", "dep:bytes"]
 compression = ["tonic/gzip"]
 tls = ["tonic/tls"]
 tls-rustls = ["dep:hyper", "dep:hyper-rustls", "dep:tower", "tower-http/util", "tower-http/add-extension", "dep:rustls-pemfile", "dep:tokio-rustls"]
 dynamic-load-balance = ["dep:tower"]
-timeout = ["tokio/time", "dep:tower"]
+timeout = ["tokio/time", "dep:tower", "tower?/timeout"]
```
Comment (djc) [inline examples/Cargo.toml:307]: Same here.

**Comment by djc** [comment]
> Allows users to use transport feature independent from channel.

Also, do you have a concrete use case in mind for this?

File: examples/Cargo.toml [hunk 2]
```diff
@@ -298,13 +298,13 @@ hyper-warp-multiplex = ["hyper-warp"]
 uds = ["tokio-stream/net", "dep:tower", "dep:hyper"]
 streaming = ["tokio-stream", "dep:h2"]
 mock = ["tokio-stream", "dep:tower"]
-tower = ["dep:hyper", "dep:tower", "dep:http"]
+tower = ["dep:hyper", "dep:tower", "tower?/timeout", "dep:http"]
```
[the remark this answers is not in the document]
Comment (djc) [reply examples/Cargo.toml:301]: (Saw this also in some of your other PRs, so would be good to clarify the goal here.)

[the remark this answers is not in the document]
Comment (tottoto) [reply examples/Cargo.toml:301]: When I tested this previously, I found that it was necessary to write it like this at the Rust version which is older than or equal to 1.70. But when I tried it again this time, it seems that it is no longer necessary. 

I don't know why the results are different even though testing is done with the same Rust version.

ref:

- https://github.com/tokio-rs/axum/pull/2350#issuecomment-1826112652
- https://github.com/tokio-rs/axum/pull/2350#issuecomment-1826115889

[the remark this answers is not in the document]
Comment (tottoto) [reply examples/Cargo.toml:301]: When I tested this previously, I found that it was necessary to write it like this at the Rust version which is older than or equal to 1.70. But when I tried it again this time, it seems that it is no longer necessary. 

I am not sure why the results are different even though testing is done with the same Rust version.

ref:

- https://github.com/tokio-rs/axum/pull/2350#issuecomment-1826112652
- https://github.com/tokio-rs/axum/pull/2350#issuecomment-1826115889

[the remark this answers is not in the document]
Comment (tottoto) [reply examples/Cargo.toml]: I think I find what happen. At this example project, changing both of them to `tower/feature` removes all `dep:tower` notation, and then it provides implicit feature. However it is hidden by the explicit `tower` feature. There is no particular effect on the example, but to unify the notation, I will use `?` in the notation.

[the remark this answers is not in the document]
Comment (djc) [reply examples/Cargo.toml]: I'm not sure I understand the full context of what you're saying. As far as I know, the `?` means that the feature dependency should not (by itself) activate the optional dependency itself. That doesn't apply here (we're actively enabling the optional feature), so we should not use the `?` here.

[the remark this answers is not in the document]
Comment (tottoto) [reply examples/Cargo.toml]: When using Rust 1.70 or older than the version, using `dep:crate-name` and `crate-name/some-feature` cannot be used at the same time because of the cargo limitation. `dep:crate-name` and `crate-name?/some-feature` is a workaround for the limitation. However, in this case, it seems to be overlooked by cargo due to the explicit feature which is the same name to the `crate-name` (`tower` in this case). Based on these premises, I prefer uniformity of notation.

[the remark this answers is not in the document]
Comment (djc) [reply examples/Cargo.toml]: So if we used 1.71 this limitation would be lifted? Sorry, I'm still not convinced.

File: tonic/Cargo.toml [hunk 3]
```diff
@@ -26,24 +26,28 @@ version = "0.11.0"
...
-  "channel",
   "dep:h2",
-  "dep:hyper", "dep:hyper-util", "dep:hyper-timeout",
+  "dep:hyper", "dep:hyper-util",
   "dep:socket2",
   "dep:tokio", "tokio?/macros", "tokio?/net", "tokio?/time",
-  "dep:tower",
+  "dep:tower", "tower?/util", "tower?/limit",
+]
+channel = [
```
Comment (djc) [inline tonic/Cargo.toml:45]: Can you update the crate root documentation to clarify a bit what the distinction is between `channel` and `transport`? I don't think the current docs are very clear:

> `transport`: Enables the fully featured, batteries included client and server implementation based on [hyper](https://docs.rs/hyper), [tower](https://docs.rs/tower) and [tokio](https://docs.rs/tokio). Enabled by default.
> `channel`: Enables just the full featured channel/client portion of the transport feature.

File: examples/Cargo.toml [hunk 4]
```diff
@@ -276,13 +276,13 @@ tracing = ["dep:tracing", "dep:tracing-subscriber"]
 uds = ["tokio-stream/net", "dep:tower", "dep:hyper", "dep:hyper-util"]
 streaming = ["tokio-stream", "dep:h2"]
 mock = ["tokio-stream", "dep:tower", "dep:hyper-util"]
-tower = ["dep:hyper", "dep:hyper-util", "dep:tower", "dep:http"]
+tower = ["dep:hyper", "dep:hyper-util", "dep:tower", "tower?/timeout", "dep:http"]
```
Comment (djc) [inline examples/Cargo.toml:279]: Does the previous commit fail to compile the examples? If so, IMO we should squash this commit into it.

File: tonic/src/transport/service/discover.rs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    hash::Hash,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::sync::mpsc::Receiver;
+
+use tokio_stream::Stream;
+use tower::discover::Change;
+
+type DiscoverResult<K, S, E> = Result<Change<K, S>, E>;
```
Comment (djc) [inline tonic/src/transport/service/discover.rs]: This seems unrelated to the commit message?

File: examples/Cargo.toml (hunk 4 above)
Comment (tottoto) [reply examples/Cargo.toml:279]: > Does the previous commit fail to compile the examples? 

Yes. Squashed these two commits.

File: examples/Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (tottoto) [reply examples/Cargo.toml]: Yes. When we bump MSRV to 1.71, this workaround is not needed.

Found the pull request which addressed this: https://github.com/rust-lang/cargo/pull/12130.

File: tonic/src/transport/service/discover.rs (hunk 5 above)
Comment (tottoto) [reply tonic/src/transport/service/discover.rs]: Fixed.

**Comment by tottoto** [comment]
Addressed the reviews, and updated the feature configs.

File: examples/Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (LucioFranco) [reply examples/Cargo.toml]: This is just for examples though so I think this is fine for the moment.
